File: android/app/src/main/java/com/vaultsync/ui/viewmodel/DocumentListViewModel.kt
package com.vaultsync.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vaultsync.domain.model.Document
import com.vaultsync.domain.model.SyncStatus
import com.vaultsync.domain.usecase.DeleteDocumentUseCase
import com.vaultsync.domain.usecase.GetDocumentsUseCase
import com.vaultsync.domain.usecase.ImportDocumentUseCase
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

@OptIn(ExperimentalCoroutinesApi::class)
class DocumentListViewModel(
    private val getDocumentsUseCase: GetDocumentsUseCase,
    private val importDocumentUseCase: ImportDocumentUseCase,
    private val deleteDocumentUseCase: DeleteDocumentUseCase
) : ViewModel() {

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _statusFilter = MutableStateFlow<SyncStatus?>(null)
    val statusFilter: StateFlow<SyncStatus?> = _statusFilter.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    val documents: StateFlow<List<Document>> = combine(_searchQuery, _statusFilter) { query, filter ->
        Pair(query, filter)
    }.flatMapLatest { (query, filter) ->
        getDocumentsUseCase(query = query, filterStatus = filter)
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = emptyList()
    )

    fun onSearchQueryChanged(query: String) {
        _searchQuery.value = query
    }

    fun onFilterChanged(status: SyncStatus?) {
        _statusFilter.value = status
    }

    fun importFile(name: String, mimeType: String, bytes: ByteArray) {
        viewModelScope.launch {
            try {
                _errorMessage.value = null
                importDocumentUseCase(name, mimeType, bytes)
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to import file"
            }
        }
    }

    fun deleteDocument(id: String) {
        viewModelScope.launch {
            try {
                _errorMessage.value = null
                deleteDocumentUseCase(id)
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to delete document"
            }
        }
    }

    fun clearError() {
        _errorMessage.value = null
    }
}

File: ios/Sources/VaultSync/Presentation/ViewModels/DocumentListViewModel.swift
import SwiftUI
import Combine

@MainActor
public final class DocumentListViewModel: ObservableObject {
    @Published public var documents: [Document] = []
    @Published public var searchQuery: String = ""
    @Published public var selectedFilter: SyncStatus?
    @Published public var errorMessage: String?

    private let getDocumentsUseCase: GetDocumentsUseCase
    private let importDocumentUseCase: ImportDocumentUseCase
    private let deleteDocumentUseCase: DeleteDocumentUseCase

    public init(
        getDocumentsUseCase: GetDocumentsUseCase,
        importDocumentUseCase: ImportDocumentUseCase,
        deleteDocumentUseCase: DeleteDocumentUseCase
    ) {
        self.getDocumentsUseCase = getDocumentsUseCase
        self.importDocumentUseCase = importDocumentUseCase
        self.deleteDocumentUseCase = deleteDocumentUseCase

        Task {
            await observeDocuments()
        }
    }

    public func observeDocuments() async {
        for await docs in getDocumentsUseCase.execute(query: searchQuery, filterStatus: selectedFilter) {
            self.documents = docs
        }
    }

    public func filterChanged(_ status: SyncStatus?) {
        selectedFilter = status
        Task {
            if let direct = try? await getDocumentsUseCase.executeDirect() {
                self.documents = direct.filter { doc in
                    let matchesQuery = searchQuery.isEmpty || doc.name.localizedCaseInsensitiveContains(searchQuery)
                    let matchesStatus = selectedFilter == nil || doc.syncStatus == selectedFilter
                    return matchesQuery && matchesStatus
                }
            }
        }
    }

    public func importFile(name: String, mimeType: String, bytes: Data) {
        Task {
            do {
                _ = try await importDocumentUseCase.execute(name: name, mimeType: mimeType, bytes: bytes)
                errorMessage = nil
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    public func deleteDocument(id: String) {
        Task {
            do {
                try await deleteDocumentUseCase.execute(documentId: id)
                errorMessage = nil
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}
